cmd/collectors/rest/rest.go

```diff
@@ -0,0 +1,214 @@
+package rest
+
+import (
+	"fmt"
+	"github.com/tidwall/gjson"
+	"goharvest2/cmd/poller/collector"
+	"goharvest2/cmd/poller/plugin"
+	"goharvest2/pkg/api/ontapi/rest"
+	"goharvest2/pkg/errors"
+	"goharvest2/pkg/matrix"
+	"strconv"
+	"time"
+)
+
+type Rest struct {
+	*collector.AbstractCollector
+	client         *rest.Client
+	apiPath        string
+	instanceKeys   []string
+	instanceLabels map[string]string
+}
+
+func init() {
+	plugin.RegisterModule(Rest{})
+}
+
+func (Rest) HarvestModule() plugin.ModuleInfo {
+	return plugin.ModuleInfo{
+		ID:  "harvest.collector.rest",
+		New: func() plugin.Module { return new(Rest) },
+	}
+}
+
+func (r *Rest) Init(a *collector.AbstractCollector) error {
+
+	var err error
+
+	r.AbstractCollector = a
+	if err = collector.Init(r); err != nil {
+		return err
+	}
+
+	if r.client, err = r.getClient(); err != nil {
+		return err
+	}
+
+	if err = r.client.Init(5); err != nil {
+		return err
+	}
+
+	r.Logger.Info().Msgf("connected to %s: %s", r.client.ClusterName(), r.client.Info())
+
+	r.Matrix.SetGlobalLabel("cluster", r.client.ClusterName())
+
+	if err = r.initCache(r.getTemplateFn(), r.client.Version()); err != nil {
+		return err
+	}
+	r.Logger.Info().Msgf("initialized cache with %d metrics", len(r.Matrix.GetMetrics()))
+	return nil
+}
+
+func (r *Rest) getClient() (*rest.Client, error) {
+	var (
+		addr, x             string
+		useInsecureTls      bool
+		certAuth, basicAuth [2]string
+	)
+
+	if addr = r.Params.GetChildContentS("addr"); addr == "" {
+		return nil, errors.New(errors.MISSING_PARAM, "addr")
+	}
+
+	if x = r.Params.GetChildContentS("use_insecure_tls"); x != "" {
+		useInsecureTls, _ = strconv.ParseBool(x)
+	}
+
+	// set authentication method
+	if r.Params.GetChildContentS("auth_style") == "certificate_auth" {
+
+		certAuth[0] = r.Params.GetChildContentS("ssl_cert")
+		certAuth[1] = r.Params.GetChildContentS("ssl_key")
+
+		return rest.New(addr, &certAuth, nil, useInsecureTls)
+	}
+
+	basicAuth[0] = r.Params.GetChildContentS("username")
+	basicAuth[1] = r.Params.GetChildContentS("password")
+
+	return rest.New(addr, nil, &basicAuth, useInsecureTls)
+}
+
+func (r *Rest) getTemplateFn() string {
+	var fn string
+	if r.Params.GetChildS("objects") != nil {
+		fn = r.Params.GetChildS("objects").GetChildContentS(r.Object)
+	}
+	return fn
+}
+
+func (r *Rest) PollData() (*matrix.Matrix, error) {
+
+	var (
+		content      []byte
+		count        uint64
+		apiD, parseD time.Duration
+		startTime    time.Time
+		err          error
+	)
+
+	r.Logger.Info().Msgf("starting data poll")
+	r.Matrix.Reset()
+
+	startTime = time.Now()
+	if content, err = r.client.Get(r.apiPath, map[string]string{"fields": "*,"}); err != nil {
```

Comment:
instead of fetching all fields, only the template required ones should be

Reply:
passing template required fields could be a problem if any of the template field is not found in rest. Rest will return error in that case. THis may happen when templates from different versions are used for target cluster.

Reply:
right, zapi ignores unknown attributes but rest does not.
Options:
- use star
- use schema to diff template with what is supported
- use start first time and diff with template on subsequent queries